src/java/com/unifina/signalpath/utils/MessageChainUtil.java

```diff
@@ -0,0 +1,73 @@
+package com.unifina.signalpath.utils;
+
+import com.streamr.client.protocol.message_layer.MessageID;
+import com.streamr.client.protocol.message_layer.MessageRef;
+import com.streamr.client.protocol.message_layer.StreamMessage;
+import com.streamr.client.protocol.message_layer.StreamMessageV30;
+import com.unifina.data.StreamPartitioner;
+import com.unifina.domain.data.Stream;
+import com.unifina.domain.security.SecUser;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MessageChainUtil implements Serializable {
+	private Map<String,Long> previousTimestamps = new HashMap<>();
+	private Map<String,Long> previousSequenceNumbers = new HashMap<>();
+
+	private transient SecUser cachedUser = null;
+
+	private static final Logger log = Logger.getLogger(MessageChainUtil.class);
+
+	private SecUser getUser(Long userId) {
+		if (cachedUser == null) {
```

Comment:
I guess we should either:
- Check that `userId` is a match with `cachedUser.getId()`, in case someone is abusing the class, or
- Even better: prevent abusing the class by passing the `userId` in the constructor of this class instead of re-passing it in `getStreamMessage` for every message, since we can't allow the user to change.